### ISSUE TITLE
Ability to create contract instance by bytecode and ACI

### ISSUE THREAD

**Issue by davidyuk** [aeternity/aepp-sdk-js#1322]
depends on: https://github.com/aeternity/aesophia_http/issues/81

**Comment by marc0olo** [comment]
> depends on: [aeternity/aesophia_http#81](https://github.com/aeternity/aesophia_http/issues/81)

why does this depend on this? I don't understand

**Comment by marc0olo** [comment]
@davidyuk I even think it should be possible to create the contract instance without relying on bytecode at all by just providing the ACI. can we do that?

if I am writing a frontend that just wants to interact with the contract (knowing the ACI) I shouldn't rely on the sourcecode or bytecode

**Comment by davidyuk** [comment]
> why does this depend on this? I don't understand

Actually, not strictly depends, but would be nice to have a confirmation of bytecode comparison approach

> I even think it should be possible to create the contract instance without relying on bytecode at all by just providing the ACI. can we do that?

Should be yes, I'll check it

> old contract id gets ditched and replaced by the contract id of the new contract deployment?

Yep, the same with `compile` method. Probably better to throw exception in these cases

**Comment by marc0olo** [comment]
> Actually, not strictly depends, but would be nice to have a confirmation of bytecode comparison approach

yeah, I think that needs more brains to figure out the best approach for this. maybe we should open a separate discussion issue for that.

> Should be yes, I'll check it

👍 

> Yep, the same with compile method. Probably better to throw exception in these cases

We should cover each use case respective. Not sure if throwing errors is the best approach. Currently I would prefer throwing an error.

What usecases do we have for the contract instance?
1. source code only (bytecode & ACI available after compile)
    - initialization of contract instance requires (implicit) compilation to obtain bytecode and ACI
    - deployment possible if `contract_id` not defined
    - deployment not possible if `contract_id` is present
2. bytecode & ACI
    - no need for communication with http compiler
    - deployment possible if `contract_id` not defined
    - deployment not possible if `contract_id` is present
3. ACI only
    - no need for communication with http compiler
    - don't allow deployments at all

do we have verification methods on the contract instance right now for validating the bytecode? if so we can only support this right now for usecase nr. 1 and we are limited to a specific compiler version the SDK is connected to, correct?

**Comment by davidyuk** [comment]
> do we have verification methods on the contract instance right now for validating the bytecode?

yep, see `validateByteCode` option: https://github.com/aeternity/aepp-sdk-js/blob/f1e9c4f514d0a294bdd0f8a7d6f5a781a32bf3d4/src/contract/aci/index.js#L99-L103

**Comment by marc0olo** [comment]
> yep, see `validateByteCode` option:

this should only be possible if source code is provided of course. should we otherwise skip with warnings or throw an error? I mean who would seriously try to validate bytecode without knowing the corresponding sourcecode? doesn't make sense to me 😅
